File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.cs
public class Solution
{
    public Dictionary<int, HashSet<int>> AdjacencyList = new Dictionary<int, HashSet<int>>();
    public IList<bool> HasApple;

    public int MinTime(int n, int[][] edges, IList<bool> hasApple)
    {
        this.HasApple = hasApple;

        foreach (var edge in edges)
        {
            if (!AdjacencyList.TryGetValue(edge[0], out var list))
            {
                list = new HashSet<int>();
                AdjacencyList[edge[0]] = list;
            }
            list.Add(edge[1]);

            if (!AdjacencyList.TryGetValue(edge[1], out list))
            {
                list = new HashSet<int>();
                AdjacencyList[edge[1]] = list;
            }
            list.Add(edge[0]);
        }

        return dfs(0, -1);
    }

    public int dfs(int node, int parent)
    {
        if (!AdjacencyList.ContainsKey(node)) return 0;

        var totalTime = 0;
        foreach (var child in AdjacencyList[node])
        {
            if (child == parent) continue;
            var childTime = dfs(child, node);
            if (childTime > 0 || HasApple[child])
                totalTime += childTime + 2;
        }
        return totalTime;
    }
}

File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        
        adjacency_list = defaultdict(set)
        
        for edge in edges:
            adjacency_list[edge[0]].add(edge[1])
            adjacency_list[edge[1]].add(edge[0])
        
        def dfs(node, parent):
            if node not in adjacency_list: return 0
            
            total_time = 0
            for child in adjacency_list[node]:
                if child == parent: continue
                child_time = dfs(child, node)
                if child_time > 0 or hasApple[child]:
                    total_time += child_time + 2
            return total_time
        
        return dfs(0, -1)
